crates/ide_assists/src/handlers/extract_struct_from_enum_variant.rs

```diff
@@ -152,48 +162,81 @@ fn existing_definition(db: &RootDatabase, variant_name: &ast::Name, variant: &Va
 
 fn create_struct_def(
     variant_name: ast::Name,
+    variant: &ast::Variant,
     field_list: &Either<ast::RecordFieldList, ast::TupleFieldList>,
     enum_: &ast::Enum,
 ) -> ast::Struct {
-    let pub_vis = make::visibility_pub();
+    let enum_vis = enum_.visibility();
 
-    let insert_pub = |node: &'_ SyntaxNode| {
-        let pub_vis = pub_vis.clone_for_update();
-        ted::insert(ted::Position::before(node), pub_vis.syntax());
+    let insert_vis = |node: &'_ SyntaxNode, vis: &'_ SyntaxNode| {
+        let vis = vis.clone_for_update();
+        ted::insert(ted::Position::before(node), vis);
     };
 
-    // for fields without any existing visibility, use pub visibility
-    let field_list = match field_list {
+    // for fields without any existing visibility, use visibility of enum
+    let field_list: ast::FieldList = match field_list {
         Either::Left(field_list) => {
             let field_list = field_list.clone_for_update();
 
-            field_list
-                .fields()
-                .filter(|field| field.visibility().is_none())
-                .filter_map(|field| field.name())
-                .for_each(|it| insert_pub(it.syntax()));
+            if let Some(vis) = &enum_vis {
+                field_list
+                    .fields()
+                    .filter(|field| field.visibility().is_none())
+                    .filter_map(|field| field.name())
+                    .for_each(|it| insert_vis(it.syntax(), vis.syntax()));
+            }
 
             field_list.into()
         }
         Either::Right(field_list) => {
             let field_list = field_list.clone_for_update();
 
-            field_list
-                .fields()
-                .filter(|field| field.visibility().is_none())
-                .filter_map(|field| field.ty())
-                .for_each(|it| insert_pub(it.syntax()));
+            if let Some(vis) = &enum_vis {
+                field_list
+                    .fields()
+                    .filter(|field| field.visibility().is_none())
+                    .filter_map(|field| field.ty())
+                    .for_each(|it| insert_vis(it.syntax(), vis.syntax()));
+            }
 
             field_list.into()
         }
     };
 
+    field_list.reindent_to(IndentLevel::single());
+
     // FIXME: This uses all the generic params of the enum, but the variant might not use all of them.
-    let strukt =
-        make::struct_(enum_.visibility(), variant_name, enum_.generic_param_list(), field_list)
-            .clone_for_update();
+    let strukt = make::struct_(enum_vis, variant_name, enum_.generic_param_list(), field_list)
+        .clone_for_update();
+
+    // FIXME: Consider making this an actual function somewhere (like in `AttrsOwnerEdit`) after some deliberation
```

Comment:
Yeah, very much this. 

I *think* we want to add this to `DocCommentsOwner` as

```rust
fn outer_attrs_and_docs(&self) -> Vec<SyntaxElement>
// alternative
fn outer_attrs_and_docs(&self) -> RangeInclusive<SyntaxElemet>
```

then `Edit` will contain logic for only actually deliting that stuff.